File: DataStructures/Map/atcoder/B54/GPT/B54.go
// 以下は **Go 1.20.6** 向けの型明示・関数化した実装です。
// 処理は **O(N)**、メモリは異なる要素の数だけ保持するため **O(U)**（U はユニーク値数）です。

// ---

// ```go
// Go 1.20.6
// 処理時間: O(N)
// メモリ使用量: O(U) (U は配列内の異なる値の数、最大で N)
package main

import (
	"bufio"
	"fmt"
	"os"
)

// countPairs は、1 ≤ j < i ≤ N かつ A_j = A_i を満たす (i, j) の組数を返します。
// N: 整数の個数
// arr: 整数スライス A1...AN
// 戻り値: 条件を満たす (i, j) の組数
func countPairs(N int, arr []int) int {
	freq := make(map[int]int) // 値 -> 出現回数
	count := 0

	for _, val := range arr {
		if c, exists := freq[val]; exists {
			// 過去の出現回数分だけ組が作れる
			count += c
			freq[val] = c + 1
		} else {
			freq[val] = 1
		}
	}
	return count
}

func main() {
	in := bufio.NewReader(os.Stdin)

	var N int
	fmt.Fscan(in, &N)

	arr := make([]int, N)
	for i := 0; i < N; i++ {
		fmt.Fscan(in, &arr[i])
	}

	result := countPairs(N, arr)
	fmt.Println(result)
}
// ```

// ---

// ### 実装のポイント

// 1. **型明示**

//    * `N int`, `arr []int`, 戻り値も `int` を明示
//    * `freq` は `map[int]int` 型
// 2. **時間計算量**

//    * 1回のスライス走査で組数を計算 → **O(N)**
//    * `map` の `exists` チェックと代入は平均 **O(1)**
// 3. **メモリ計算量**

//    * 最大で `map` に N エントリ（全要素が異なる場合）
//    * N=100,000 程度なら数MB程度の使用

File: DataStructures/Map/atcoder/B54/GPT/B54.py
# 以下は **Python (CPython 3.11.4)** 向けの実装で、型アノテーションを明示し、処理時間・メモリを意識した **O(N)** 解法です。
# `dict` を使って値の出現回数を記録し、1回の走査で計算します。

# ---

# ```python
# Python 3.11.4
# 処理時間: O(N)
# メモリ使用量: O(U) (U は配列内の異なる値の数、最大で N)

from sys import stdin
from typing import List


def count_pairs(N: int, arr: List[int]) -> int:
    """
    条件を満たす (i, j) の組数をカウントする

    Parameters
    ----------
    N : int
        配列の要素数
    arr : List[int]
        長さ N の整数配列 A1...AN

    Returns
    -------
    int
        条件 (1 ≤ j < i ≤ N かつ A_j = A_i) を満たす組数
    """
    freq: dict[int, int] = {}  # 値 -> 出現回数
    count: int = 0

    for val in arr:
        if val in freq:
            # 過去に出現した回数分だけ組を作れる
            count += freq[val]
            freq[val] += 1
        else:
            freq[val] = 1
    return count


# 高速な入力処理
data: List[int] = list(map(int, stdin.read().strip().split()))
N: int = data[0]
arr: List[int] = data[1:]

result: int = count_pairs(N, arr)
print(result)
# ### 実装のポイント

# 1. **型アノテーション**

#    * `N: int`, `arr: List[int]`, `freq: dict[int, int]`, `count: int` を明示
#    * Pylance で型エラーが出ないように、変数初期化時に型を確定
# 2. **時間計算量**

#    * 1回のループ (O(N)) で組数計算
#    * `dict` の `in`, `[]` アクセスは平均 O(1)
# 3. **メモリ計算量**

#    * `freq` は異なる値の数だけエントリ保持（最大 N）
#    * N=100,000 でも数MB程度

# ---

# このままではテキスト出力ですが、希望があれば
# 処理の各ステップを図解した **Markdown解説** も Python 用に作成できます。
